pkg/solana/logpoller/parser.go

```diff
@@ -0,0 +1,194 @@
+package logpoller
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+type MessageStyle string
+
+const (
+	MessageStyleMuted   MessageStyle = "muted"
+	MessageStyleInfo    MessageStyle = "info"
+	MessageStyleSuccess MessageStyle = "success"
+	MessageStyleWarning MessageStyle = "warning"
+)
+
+var (
+	invokeMatcher      = regexp.MustCompile(`Program (\w*) invoke \[(\d)\]`)
+	consumedMatcher    = regexp.MustCompile(`Program \w* consumed (\d*) (.*)`)
+	logMatcher         = regexp.MustCompile(`Program log: (.*)`)
+	dataMatcher        = regexp.MustCompile(`Program data: (.*)`)
+	instructionMatcher = regexp.MustCompile(`Instruction: (.*)`)
+)
+
+type BlockData struct {
+	BlockNumber         uint64
+	BlockHash           string
+	TransactionHash     string
+	TransactionIndex    int
+	TransactionLogIndex uint
+}
+
+type ProgramLog struct {
+	BlockData
+	Text   string
+	Prefix string
+	Style  MessageStyle
+}
+
+type ProgramEvent struct {
+	BlockData
+	Prefix       string
+	FunctionName string
+	Data         string
+}
+
+type ProgramOutput struct {
+	Program      string
+	Logs         []ProgramLog
+	Events       []ProgramEvent
+	ComputeUnits uint
+	Truncated    bool
+	Failed       bool
+}
+
+func prefixBuilder(depth int) string {
+	return strings.Repeat(">", depth)
+}
+
+/*
+Program J1zQwrBNBngz26jRPNWsUSZMHJwBwpkoDitXRV95LdK4 invoke [1]
+Program log: Instruction: CreateLog
+Program data: HDQnaQjSWwkNAAAASGVsbG8sIFdvcmxkISoAAAAAAAAA // base64 encoded; borsh encoded with identifier
+Program J1zQwrBNBngz26jRPNWsUSZMHJwBwpkoDitXRV95LdK4 consumed 1477 of 200000 compute units
+Program J1zQwrBNBngz26jRPNWsUSZMHJwBwpkoDitXRV95LdK4 success
+*/
+func parseProgramLogs(logs []string) []ProgramOutput {
+	var depth int
+
+	instLogs := []ProgramOutput{}
+	lastEventIdx := -1
+
+	for _, log := range logs {
+		if strings.HasPrefix(log, "Program log:") {
+			logDataMatches := logMatcher.FindStringSubmatch(log)
+
+			if len(logDataMatches) <= 1 {
+				continue
+			}
+
+			instructionMatches := instructionMatcher.FindStringSubmatch(logDataMatches[1])
+
+			if len(instructionMatches) > 1 {
+				// is an event which should be followed by: Program data: (.*)
+				instLogs[len(instLogs)-1].Events = append(instLogs[len(instLogs)-1].Events, ProgramEvent{
```

Comment:
It can if the program log output for solana changes. Program output always starts with an invoke message which is handled. The other way to do this would be to track a `lastLogIndex` value. It would be safer.